addon/services/modal.ts

```diff
@@ -0,0 +1,95 @@
+import Service from '@ember/service';
+import Evented from '@ember/object/evented';
+import { A } from '@ember/array';
+import { set } from '@ember/object';
+import { later } from '@ember/runloop';
+import { notEmpty } from '@ember-decorators/object/computed';
+
+interface ModalConfig {
+    outlet: string | undefined;
+}
+
+/**
+ * Class that handles opening and closing of modals
+ *
+ * @class modal
+ * @extends Ember.Service
+ */
+
+export default class Modal extends Service.extend(Evented) {
+    animation: string | undefined;
+    animationIn: string = 'zoomIn';
+    animationOut: string = 'zoomOut';
+    animationDuration: number = 500;
+    current: any;
+    modals = A();
+    outlets = A();
+
+    @notEmpty('current') modalIsOpen!: boolean;
```

Comment:
what do the `!`'s signify in typescript?

Reply:
They are a "definite" assertion, which means that a variable is indeed assigned for all intents and purposes, even if TypeScript’s analyses cannot detect so. This is usually due to dependency injection